### ISSUE TITLE
Toward a Bazel build system

### ISSUE THREAD

**Issue by themarwhal** [magma/magma#8338]
# Toward a Bazel build system

The Magma project desires an improved, unified build system. This proposal outlines an experimental effort in bazel-ifying SessionD, as an initial proof of concept.

## Motivation

Our C++ build system needs revamping in order to improve developer efficiency.  

Some of the major pain points are

1. Long build times across VM restarts
    1. MME fresh build takes ~30 minutes
2. Builds are occasionally unreliable
    1.  `make clean` is occasionally needed to get the correct build
3. It is very difficult for non-experts to make changes to our builds
    1. Even things as simple as adding new unit tests (a major team need)
    2. It is very challenging to self-educate on Magma CMake using public CMake documentation

Whether we stick with CMake or migrate to a new build system like Bazel, we will need a major rework to make the best use out of developers’ time.

 I’ve added a section below on some advantages for both CMake and Bazel, but this effort will focus on the Bazel approach.

## Goal

The goal of this effort is to bazel-ify SessionD as a proof-of-concept, without affecting any existing build system. 

Once we have a build, we can measure any build time improvements, reliability improvements, and feasibility and effort level of migrating MME and the rest of Magma. Additionally, we will verify that Bazel built SessionD is compatible with existing tests and dev environments (e.g. IntelliJ).

Another key point to look out for during this effort is Bazel’s compatibility with various code generation we have to do in MME (e.g. ASN.1). Bazel supports [genrules (General Rules)](https://docs.bazel.build/versions/main/be/general.html) that provides a way to generate files via bash scripts, so that could be useful here.

If we decide against Bazel after the POC, undoing the work will be a matter of deleting all BUILD.bazel files from the repository.

## Non-Goals

* This effort will *NOT* affect any production or development builds that we currently rely on
* This effort does *NOT* include bazel-ification of any other service other than SessionD 

### Advantages of sticking with CMake

1. Existing developer environments work with CMake
    1. For example, running SessionD with break points is fully supported with the current build in IntelliJ
    	1. However, it is important to note that we faced some difficulty during a major CMake refactor without breaking this (See https://github.com/magma/magma/pull/6423). And CLion's remote debug with breakpoints should work as long as there is an executable.
    
### Advantages of migrating to Bazel

1. Potential build time improvements from better cache support
2. Clear documentation on Bazel best practices
    1. See  https://docs.bazel.build/versions/main/best-practices.html
    2. Bazel requires explicit dependency declarations
    3. Bazel does not support circular dependencies
    4. [Buildifier](https://github.com/bazelbuild/buildtools) is a well supported formatter for all BUILD files
3. Easy to do partial builds
    1. Since Bazel forces you to define multiple small libraries, it’s very easy to rebuild only the components you touch
4. Free improvements in testing tools
    1. Test groupings
    2. Test timeouts
    3. Clear success / failure enumeration - can be used to generate XML summaries to upload in CI
5. Has potential to become a single build system for all of Magma
    1. This aligns better with the monorepo philosophy
    2. Multi-language support - Python, Golang, etc.
6. IDE support to be tested
	1. [Bazel integration for CLion](https://plugins.jetbrains.com/plugin/9554-bazel)
	2. [Bazel integration for VSCode](https://marketplace.visualstudio.com/items?itemName=BazelBuild.vscode-bazel)

### Why not Buck?

1. Buck and Bazel are very similar but Bazel has more opensource support
    1. [Bazel has 17K stars](https://github.com/bazelbuild/bazel) on GitHub vs [8K for Buck](https://github.com/facebook/buck)
    2. Any upstream changes for Buck have to go through FB internal, whereas Bazel is community pushed (This can easily be seen in their commit histories)
2. A lot of our key dependencies come with Bazel support but not Buck - protobuf, etc.
3. Buck is lacking in Go support compared to Bazel
    1. [Uber migrated from Buck to Bazel for this reason](https://eng.uber.com/go-monorepo-bazel/)
4. @amarpad [mentioned in the comments](https://github.com/magma/magma/issues/8338#issuecomment-889451817) that we already had to switch away from Buck when we went open source

**Comment by ulaskozat** [comment]
Looks like a lot of positive points to gain. Isn't there a good support for dev environment for Bazel; i.e., is the advantage of CMake listed on this front a real advantage?

**Comment by markjen** [comment]
> Looks like a lot of positive points to gain. Isn't there a good support for dev environment for Bazel; i.e., is the advantage of CMake listed on this front a real advantage?

There is a bazel plugin for CLion/IntelliJ. We should test it out and see if it works out of the box.

Also, the CMake integration with CLion is not good (e.g. attempting to load lte/gateway/c/core/CMakeLists.txt doesn’t work well for me at least, it has trouble with ExternalProject), so perhaps this may end in a tie :)

**Comment by amarpad** [comment]
Three notes:
1 - The one thing I would also suggest is that while we Bazelify things we don't make the same build mistakes (outside the build system) like having 6 different sub libraries in common etc. So orthogonal to this proposal it might be prudent to have some refactor proposal when we are touching the makefiles for specific components.

2 - Originally we used to build with Buck but a lot of Buck features that are available in FB are not available in opensource as I think the project is not actively maintained outside and so we ended up flipping to CMake. This should also serve as a datapoint to not use buck.

3 - Just think through if there are any implications on OAI CUPS. I don't see any as long as you have an MME only target with some additional options, but might be worth thinking through.

**Comment by themarwhal** [comment]
> Three notes:
> 1 - The one thing I would also suggest is that while we Bazelify things we don't make the same build mistakes (outside the build system) like having 6 different sub libraries in common etc. So orthogonal to this proposal it might be prudent to have some refactor proposal when we are touching the makefiles for specific components.
> 
> 2 - Originally we used to build with Buck but a lot of Buck features that are available in FB are not available in opensource as I think the project is not actively maintained outside and so we ended up flipping to CMake. This should also serve as a datapoint to not use buck.
> 
> 3 - Just think through if there are any implications on OAI CUPS. I don't see any as long as you have an MME only target with some additional options, but might be worth thinking through.

1. Yes, I think this is slightly orthogonal but Bazel in general promotes a flatter directory structure which aligns with your point. :)
2. That is a good input. I'll add it to the list in 'Why not Buck?'
3. Not very familiar on OAI CUPS :D Do you mean the various different feature flags that we support in the Makefile?

**Comment by hcgatewood** [comment]
Accepted. Keep this issue open until the PoC is merged, for discussing the PoC's improvements/affordances. Once we have clear metrics/feel, let's close the issue

**Comment by themarwhal** [comment]
# Summary Of Effort

## Accomplished goals

- [X] Bazelify SessionD executable and all unit tests in `lte/gateway/c/session_manager` and `orc8r/gateway/c/`
- [X] Validate the build by running through all integration tests, teravm tests, existing performance tests
- [X] Establish a VSCode based workflow that integrates with Bazel (Docs: [VSCode+Bazel](https://github.com/magma/magma/blob/master/experimental/bazel-base/VSCodeBazel.md), Slides: [Bazel Slides](https://docs.google.com/presentation/d/1V1LRI7yfxRzdQ9a7Hg_WvFB-GMkv2Cofs85u5uFTLv4/edit))


## Some stats

1. Building SessionD from scratch with all external dependencies takes ~7 min
2. With disk cache placed on the host machine, rebuilding SessionD on a fresh VM goes down to ~1 min. (The external dependencies are often built only once as it is cached.)
3. Re-running all SessionD unit tests after some modification is faster with Bazel by around 1 minute. (CMake: 3m23.382s, Bazel: 2m24.239s) (Doc: https://docs.google.com/document/d/10RsTMwapMkNl5JEKYJ_Dj69lr0G4k6do8bormmFb-PM/edit?usp=sharing)

## Future work
1. Migrate code coverage generation to be done with Bazel
2. Migrate SessionD to bazel and remove CMake
3. Investigate best caching mechanism for CI and dev workflow

**Comment by mattymo** [comment]
My concern with Bazel is that it doesn't support Dockerfiles. You need to rewrite all Dockerfiles as rules_docker. https://github.com/bazelbuild/rules_docker

This dramatically increases disk space usage and build time. Building anything else that isn't inside a Docker image is just fine with Bazel.

**Comment by electronjoe** [comment]
@mattymo I think the plan (@themarwhal can confirm) might be to:

1. Initially use Bazel to build artifacts inside Dockerfiles - as an alternative to python + Makefile + Cmake mixtures today.
2. As we migrate towards containerized AGW we use `rules_docker` to build these containers.  At this point the docker containers will be **very small** - we will have removed shared library use and so the containers will be ~{binaries, config files} and nothing more.

I think at phase 2 Bazel build of Docker containers (well, they need not be Docker containers - but generalized containers conformant to the new container APIs) will take ~seconds beyond the actual build time of the artifacts, and will generate containers that are ~100 MB TOPS (mme) not GBs as today.

This applies to Python as well, in which PAR files would be generated by Bazel bundling all deps. Mitigating the python environment dependency hell we have today.

That would be the general shape of the strategy - I suspect.

But note that phase 2 is purely optional (I don't know why you wouldn't do it though) - and that bridge can be crossed in a separate proposal when we get there.

**Comment by themarwhal** [comment]
On top of what @electronjoe has already added, we can even build the necessary executables outside so that they can be pulled into the image.

**Comment by hcgatewood** [comment]
Note: officially passed by TSC. Bazel is now the planned path forward for Magma. Thanks @themarwhal for all your great work!
